### ISSUE TITLE
Taxon-specific form for butterfly recording for use in activities

### ISSUE THREAD

**Issue by DavidRoy** [BiologicalRecordsCentre/iRecord#1073]
Requirement for a taxon-specific recording form for butterflies:
1. List-based input form to match use within the 'General survey' mode within the iRecord app
2. Add Butterfly Conservation logo
2. Use dynamic attributes already setup for butterflies
3. Add form to iRecord menu - Record-Species group forms
4. Enable form to be available to activities
5. Enable form to be available for use within the iRecord app

@johnvanbreda can you pick this up under recent work for Defra as needed by end of June at latest

**Comment by johnvanbreda** [comment]
@DavidRoy what do you mean specifically by "match use within the 'General survey' mode within the iRecord app"? The app is quite different to a standard list of records form because it captures a separate precise location in a sub-sample for every record. One option is to add a grid reference column to the species entry grid but I'm not sure that's what you are expecting?

**Comment by johnvanbreda** [comment]
@DavidRoy the requirements changed after our Teams discussion to just needing the ability to use the standard multisite form for activities. I've done that – you can add “Enter records at several places on one date” to any activity.

Is that enough or do you still need a butterfly specific version of this form?

**Comment by DavidRoy** [comment]
Thanks.  Is another change to make for the app to use this form as 'input_form' for records from the general survey mode in the app?

**Comment by johnvanbreda** [comment]
The app already uses a similar form (https://www.brc.ac.uk/irecord/enter-app-record-list) though I note it doesn't seem to have occurrence attributes when you edit the data. Let me know if you want me to look into this @DavidRoy.

**Comment by DavidRoy** [comment]
Yes please, can you update the form to enable editing of occurrence attributes (dynamic across taxon groups?)

**Comment by johnvanbreda** [comment]
@kazlauskis The problem here is the edit form associated with the app General Survey function does not allow the user to edit any of the occurrence attributes. This is because the only occurrence attribute linked to the survey (id 576) is breeding evidence (birds) which is ID 823. However if I run a query to scan a data for the different attribute values associated with this survey I find the following:
2	"Abundance Dafor"
16	"Abundance"
18	"Identified By"
34	"Ad"
35	"Co"
36	"Ov"
39	"Em"
105	"Sex"
106	"Stage"
203	"Butterfly app abundance code"
293	"Butterfly life stage"
523	"Abundance"
823	"Breeding evidence (birds)"

Is there any reason these aren't linked to the survey? Also - is this a complete list or does the app include any other occurence attribute in the General Survey?

**Comment by kazlauskis** [comment]
@johnvanbreda I have now added the missing attributes to the general survey. Here is the full list of the occurrence attributes: https://warehouse1.indicia.org.uk/index.php/attribute_by_survey/576?type=occurrence

**Comment by johnvanbreda** [comment]
@kazlauskis thanks. We also need to link these attributes to the appropriate high-level taxon they appear for. I can probably work this out but it would be good if the behaviour is exactly like the app - can you let me know how the app decides when to display each attribute please?

**Comment by kazlauskis** [comment]
@kitenetter It would be good to have a single doc that would work as an interface between the website and the app (links to NERC-CEH/irecord-app/issues/211). 

@johnvanbreda These are the taxon groups currently in use:

name | groups | occAttrs | smpAttrs
-- | -- | -- | --
birds | [73] | [823, 16, 523, 106, 105, 18] | &nbsp;
bryophytes | [133, 129] | [470, 471, 475, 476, 472, 473, 474] | [208]
butterflies | [104] | [105, 293, 203, 18] | &nbsp;
dragonflies | [107] | [34, 35, 36, 37, 38, 39, 842] | [59]
plants-fungi | [89, 78, 87, 99, 81, 148, 92] | [16, 523, 2, 105, 106, 18] | &nbsp;

**Comment by DavidRoy** [comment]
@johnvanbreda is this now resolved?

**Comment by johnvanbreda** [comment]
@kazlauskis - the survey includes occurrence attribute ID 2 (Abundance Dafor) - should this be linked to Plantae? Or is it not used?

Note that the taxon linking of attributes in the website forms is done via higher taxa, not groups. I've linked the attributes in the same way as far as I can but there may be some oddities due to this.
Birds - mapped to Aves
Bryophytes - mapped to Bryophyta
Butterflies - mapped to Papilionoidea
Dragonflies - mapped to Odonata
plants-fungi - mapped to Plantae, Fungi, Chromista

@DavidRoy - was just typing this out as you posted!

Even having done that, there are quite a few issues with the way this form works compared with the app. The dynamic attribute handling in a grid form is based on the premise that an undefined taxon (i.e. a grid row before you enter a species name) has some attributes, each of which create a column in the grid with a specific "system function", e.g. sex, stage, abundance. Then, when a specific taxon is chosen, this column's attribute cell get's replaced with the correct attribute for the taxon, if there is a taxon-restricted attribute with the same system function. In list entry mode, this approach is essential to avoid the grid changing the number of columns on every row.

In order to facilitate the above, I've removed the taxon restriction from the Identified By attribute (18), Abundance (16) Sex (105) and Stage (106). These generic attributes create placeholder columns which can then be filled in with the dynamic attributes when a taxon is chosen.

However, having a single column for Sex is not compatible with the way the app is capturing Sex for bryophytes, since the app has different attributes for male and female. Similarly the Stage attributes for Odonata aren't compatible with this approach.

Writing a data entry form that behaves like the app would be very tricky. Another alternative would be to use a single record form to handle the app general survey records as they don't have the restrictions of a grid column layout.

**Comment by DavidRoy** [comment]
@johnvanbreda thanks for the update.
A can of worms (or bryophytes?) has been opened.  Can you resolve how butterflies work between the app and website forms to close off this issue (simpler because the taxon group makes sense taxonomically).  Then open a new issue for the wider complications of groups-higher taxonomy-system functions?

**Comment by kazlauskis** [comment]
@johnvanbreda Yes, please link it too. The Dafor attribute isn't implemented in the single-occurrence survey yet but it is used in the general list survey in the app - this attribute will be added in the future but since you're working on it then it makes sense to link it to the Plantae now.
